File: backend/src/main/java/tom/projects/monkeylog/service/ExerciseGroupService.java
package tom.projects.monkeylog.service;

import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;
import tom.projects.monkeylog.dto.workout.ExerciseGroupCreateRequest;
import tom.projects.monkeylog.model.workout.ExerciseGroup;
import tom.projects.monkeylog.model.workout.ExerciseRow;
import tom.projects.monkeylog.model.workout.Workout;
import tom.projects.monkeylog.model.workout.WorkoutType;
import tom.projects.monkeylog.repository.workout.ExerciseGroupRepository;
import tom.projects.monkeylog.repository.workout.WorkoutRepository;
import tom.projects.monkeylog.security.AuthenticatedUser;

import java.util.UUID;

@RequiredArgsConstructor
@Service
public class ExerciseGroupService {
    private final ExerciseService exerciseService;
    private final ExerciseGroupRepository exerciseGroupRepository;
    private final WorkoutRepository workoutRepository;
    private final WorkoutService workoutService;

    public ExerciseGroup get(UUID id) {
        return exerciseGroupRepository.findById(id)
                .filter(AuthenticatedUser::isResourceOwner)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Exercise group not found"));
    }

    public ExerciseGroup addRow(UUID id) {
        ExerciseGroup exerciseGroup = get(id);
        ExerciseRow exerciseRow = ExerciseRow
                .clone(exerciseGroup.getExerciseRows().get(exerciseGroup.getExerciseRows().size() - 1));
        exerciseRow.setIsLifted(exerciseGroup.getWorkout().getWorkoutType() == WorkoutType.COMPLETED);
        exerciseGroup.addExerciseRow(exerciseRow);

        return exerciseGroupRepository.save(exerciseGroup);
    }

    public void save(ExerciseGroupCreateRequest exerciseGroupCreateRequest, UUID workoutId) {
        Workout workout = workoutService.getWorkout(workoutId);

        exerciseService.allById(exerciseGroupCreateRequest.getExerciseIds())
                .map(exercise -> {
                    ExerciseGroup exerciseGroup = new ExerciseGroup();
                    exerciseGroup.setExercise(exercise);
                    exerciseGroup.setWorkout(workout);
                    exerciseGroup.addExerciseRow(ExerciseRow.first(exerciseGroup, workout.getWorkoutType() == WorkoutType.COMPLETED));

                    return exerciseGroup;
                }).forEach(workout::addExerciseGroup);

        workoutRepository.save(workout);
    }

    public void delete(UUID id) {
        ExerciseGroup exerciseGroup = get(id);
        Workout workout = exerciseGroup.getWorkout();
        workout.getExerciseGroups().remove(exerciseGroup);
        workoutRepository.save(workout);
    }
}

File: backend/src/main/kotlin/tom/com/monkeylog/service/ExerciseGroupService.kt
package tom.com.monkeylog.service

import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.web.server.ResponseStatusException
import tom.com.monkeylog.common.dto.IdProjection
import tom.com.monkeylog.dto.workout.ExerciseGroupCreateRequest
import tom.com.monkeylog.model.workout.ExerciseGroup
import tom.com.monkeylog.model.workout.ExerciseRow
import tom.com.monkeylog.model.workout.Workout
import tom.com.monkeylog.model.workout.WorkoutType
import tom.com.monkeylog.repository.ExerciseGroupRepository
import tom.com.monkeylog.repository.WorkoutRepository
import tom.com.monkeylog.security.AuthenticatedUser
import java.util.*

@Service
class ExerciseGroupService(
    private val exerciseGroupRepository: ExerciseGroupRepository,
    private val workoutRepository: WorkoutRepository,
    private val workoutService: WorkoutService,
    private val exerciseService: ExerciseService,
) {
    fun get(id: UUID): ExerciseGroup = exerciseGroupRepository.findById(id)
        .filter(AuthenticatedUser::isResourceOwner)
        .orElseThrow { ResponseStatusException(HttpStatus.NOT_FOUND, "Exercise group not found") }


    fun addRow(id: UUID): ExerciseGroup {
        return get(id).run {
            val exerciseRow: ExerciseRow = exerciseRows.last().clone(this)
            exerciseRow.lifted = workout.workoutType === WorkoutType.COMPLETED
            exerciseRows.add(exerciseRow)

            exerciseGroupRepository.save(this)
        }
    }

    fun save(exerciseGroupCreateRequest: ExerciseGroupCreateRequest, workoutId: UUID) {
        val workout: Workout = workoutService.getWorkout(workoutId)

        exerciseService.allExercisesById(exerciseGroupCreateRequest.exerciseIds)
            .map { exercise ->
                ExerciseGroup(
                    userId = AuthenticatedUser.id,
                    exercise = exercise,
                    workout = workout
                ).apply {
                    exerciseRows.add(
                        ExerciseRow(
                            exerciseGroup = this,
                            userId = AuthenticatedUser.id,
                            lifted = workout.workoutType === WorkoutType.COMPLETED,
                        )
                    )
                }
            }.forEach(workout.exerciseGroups::add)

        workoutRepository.save(workout)
    }

    fun delete(id: UUID) {
        get(id).run {
            workout.exerciseGroups.remove(this);
            workoutRepository.save(workout)
        }
    }

    fun getExerciseGroupsByExerciseId(exerciseId: UUID, pageable: Pageable): Page<IdProjection> {
        return exerciseGroupRepository.findAllByExerciseIdAndUserIdAndWorkoutWorkoutType(
            exerciseId,
            AuthenticatedUser.id,
            WorkoutType.COMPLETED,
            pageable
        )
    }

    fun getExerciseGroups(ids: List<UUID>): List<ExerciseGroup> {
        return exerciseGroupRepository.findAllByIdIn(ids)
    }
}
